adam/visualization/positioning.py

```diff
@@ -44,99 +46,195 @@ def create_at_random_position(
+        return diff / torch.norm(diff)
+
+    def up_face(self):
+        diff = self.up_corner() - self.zero_corner()
+        return diff / torch.norm(diff)
+
+    def get_face_norms(self) -> torch.Tensor:
+        # in axis-aligned case these are always the same
+        return torch.stack([
+            self.right_face(),
+            self.forward_face(),
+            self.up_face()
+        ])
 
+    @staticmethod
+    def get_min_max_projections(projections: torch.Tensor):
+        """
 
-class PositionSolver:
-    # constants to eventually use
-    POSITION_CONSTRAINTS = {
-        "X_RANGE": (-10.0, 10.0),
-        "Y_RANGE": (-10.0, 5.0),
-        "Z_RANGE_GROUND": (0.0, 0.0),
-        "Z_RANGE_FLY": (2.0, 5.0),
-        "Z_RANGE_OTHER": (0.0, 3.0),
-    }
+        Args:
+            projections: (3, 8) tensor -> corner projections onto each of three dimensions
 
-    CLOSE_DISTANCE = 0.25
+        Returns:
+            (3, 2) tensor -> (min, max) values for each of three dimensions
 
-    def __init__(self, main_point: ndarray, points: List[ndarray]) -> None:
-        self.main_position = main_point
-        self.positions = points
+        """
+        return torch.tensor([
+            [torch.min(projections[i]), torch.max(projections[i])] for i in range(3)
+        ], dtype=torch.float)
 
+    def get_projections(self, axes: torch.Tensor):
+        """
 
-class CollisionPenalty(nn.Module):
-    def __init__(self):
-        super().__init__()
+        Args:
+            axes: (3,3) tensor -> the three axes we are projecting points onto
 
-    def forward(self, bounding_box_1: AxisAlignedBoundingBox,
-                bounding_box_2: AxisAlignedBoundingBox) -> torch.Tensor:
-        # get all corners for current center
-        corners = torch.ones((8, 3)) * self.center + self.corner_matrix
-        print(f"corners.shape: {corners.shape}")
-        # project corners onto faces from static bb (output shape (3, 8, 1) )
-
-        projections = torch.tensor(
-            [corners[j].dot(self.static_bb_faces[i]) for i in range(3) for j in range(8)],
-            dtype=torch.float,
-        ).reshape((3, 8, 1))
-        print(f"corner projections shape: {projections.shape}")
-        print(f"corner projections: {projections}")
-
-        face_min_max = torch.tensor(
-            [[torch.min(projections[i]), torch.max(projections[i])] for i in range(3)],
-            dtype=torch.float,
-        )
-        print(f"face min-maxes shape: {face_min_max.shape}")  # (3, 2)
-        print(f"face min-maxes: {face_min_max}")
+        Returns:
+            (3, 8) tensor -> each point projected onto each of three dimensions
 
-        # compute overlap between the parameter's min/maxes and the static ones
+        """
+        check_arg(axes.shape == (3,3))
+        corners = self.get_corners()
+        return torch.tensor([
+            corners[j].dot(axes[i]) for i in range(3) for j in range(8)
```

Comment:
Turns out I was wrong about which section is tricky! I need to get rid of the _iterating over one tensor to produce another_ code in three of my functions s/t a gradient can be computed all the way back to `AxisAlignedBoundingBox.center`